pombola/south_africa/views/download.py

```diff
@@ -0,0 +1,105 @@
+import os
+
+from django.db.models import Prefetch, Q
+from django.http import StreamingHttpResponse
+from django.views.generic import TemplateView
+
+import xlsx_streaming
+from pombola.core.models import Contact, Organisation, Person, Position
+
+MP_DOWNLOAD_TEMPLATE_SHEET = os.path.join(
+    os.path.dirname(os.path.realpath(__file__)), "mp-download-template.xlsx"
+)
+
+
+def generate_sheet_name(house=None):
+    """
+    Generate the sheet name of the file that will be downloaded using the house.
+    """
+    if house and house != "all":
+        return house + "-members"
+    return "members-of-parliament"
+
+
+def get_organisation_query_filter_for_house(house="all"):
+    """
+    Get the query to filter the organisations by given the house.
+    """
+    ncop_query = Q(organisation__kind__slug="provincial-legislature")
+    na_query = Q(organisation__slug="national-assembly")
+
+    if house == "ncop":
+        return ncop_query
+    elif house == "national-assembly":
+        return na_query
+    else:
+        return ncop_query | na_query
+
+
+def download_members_xlsx(request):
+    house_param = request.GET.get("house", "all")
+    house_query = get_organisation_query_filter_for_house(house_param)
+
+    # Get all of the currently_active positions at the houses
+    house_positions = (
+        Position.objects.currently_active().filter(house_query).values("id")
+    )
+
+    party_positions = (
+        Position.objects.currently_active()
+        .filter(title__slug="member")
+        .filter(organisation__kind__slug="party")
+        .select_related("organisation")
+    )
+
+    cell_phone_contacts = Contact.contact_number_contacts()
+    email_contacts = Contact.email_contacts()
+
+    # Get the persons from the positions
+    persons = (
+        Person.objects.filter(hidden=False)
+        .filter(position__id__in=house_positions)
+        .distinct()
+        .prefetch_related(
+            "alternative_names",
+            Prefetch(
+                "position_set",
+                queryset=party_positions,
+                to_attr="active_party_positions",
+            ),
+            Prefetch("contacts", queryset=cell_phone_contacts, to_attr="cell_numbers"),
+            Prefetch("contacts", queryset=email_contacts, to_attr="email_addresses"),
+        )
+    )
+
+    def get_email_address_for_person(person):
+        if person.email:
+            return person.email
+        if len(person.email_addresses) > 0:
+            return person.email_addresses[0].value
+        return ""
+
+    def yield_people():
```

Comment:
how about calling this a generator?
Something like

```suggestion
    def person_row_generator():
```

Reply:
how about pulling this out of the class and making it a function? then you can unit test it